### ISSUE TITLE
Android: Switch databases with biometric protection not working

### ISSUE THREAD

**Issue by Trekky12** [authpass/authpass#153]
Hey, 
so I recently discovered this app and I very impressed by the development.
Unfortunately I noticed that it is not possible to switch between multiple databases with biometric protection enabled.

### Steps to reproduce

1. Open the first database, insert the master password and enable biometric protection
2. Lock the database
3. It is now only possible to "Unlock Files" or "Close password files". Wheter I click "Unlock Files" or "Close password files" the biometric password prompt is shown and when authenticated the same database is opened. When the biometric prompt is canceled another database can be opened.
4. When the biometric prompt is canceled and a second database (or the first database) is selected the master password needs to be inserted.

Also notice than when the master password is not inserted and the app is closed and then opened, the previous opened database (which has a prompt for the master password) is again available with biometric authentication. With this "workaround" is is possible to open the previous database without retyping the master password, but unfortunately it is not possible to open the second database with this workaround.

So is it possible to save the biometric state for individual databases and get the possibility to easily switch between databases whithout inserting the master password each time?

**Comment by hpoul** [comment]
Thanks for the feedback. I guess this is a bit of a naming issue. 🤔️ This is how it works right now, ignore the meaning of the words 😅️:

* **Lock Files**: Technically this "closes" all files (ie. removes them from memory completely) but keeps the hash of the masterkey available in the keychain (iOS, macOS), [keystore  (android)](https://developer.android.com/training/articles/keystore) protected by biometric authentication (using platform encryption).
* **Close Files** (on the "Lock" screen): Removes the hashes from the keychain/keystore.

I hope this makes sense?

Also, have you tried to simply open bot files, both files will be able to be reopened through the biometric authentication afterwards.

If you don't want to have your master password hash stored in the keychain/keystore deactivate the checkbox. (Issue #142 tracks that the state of this checkbox is persisted, so you wouldn't have to uncheck it every time).

To sum up how I thought about it: 

* If you want a "quick unlock" for both files.. just open both files at the same time.
* If you don't want to save master password hash in the keychain/keyring, uncheck the 'save password' checkmark.

What do you think?

**Comment by Trekky12** [comment]
> 
> 
> Thanks for the feedback. I guess this is a bit of a naming issue. 🤔️ This is how it works right now, ignore the meaning of the words 😅️:
> 
> * **Lock Files**: Technically this "closes" all files (ie. removes them from memory completely) but keeps the hash of the masterkey available in the keychain (iOS, macOS), [keystore  (android)](https://developer.android.com/training/articles/keystore) protected by biometric authentication (using platform encryption).
> 
> * **Close Files** (on the "Lock" screen): Removes the hashes from the keychain/keystore.
> 
> 
> I hope this makes sense?

Actually I can't reproduce this. No mather which button I press the biometric authentication prompt is shown afterwards and the last database is loaded after authentication (without prompt for master password)
> 
> Also, have you tried to simply open both files, both files will be able to be reopened through the biometric authentication afterwards.
> 
> If you don't want to have your master password hash stored in the keychain/keystore deactivate the checkbox. (Issue #142 tracks that the state of this checkbox is persisted, so you wouldn't have to uncheck it every time).
> 
> 
> To sum up how I thought about it:
> 
>* If you want a "quick unlock" for both files.. just open both files at the same time.

When opening both files at the same time this is indeed possible, but it is not that convenient when all passwords of all databases are available. I would prefer to have them separate. Is this possible?

> 
> * If you don't want to save master password hash in the keychain/keyring, uncheck the 'save password' checkmark.
> 

Actually saving the master password hash is okay for me if this means that the biometric authentication is working and

**Comment by hpoul** [comment]
> Actually I can't reproduce this. No mather which button I press the biometric authentication prompt is shown afterwards and the last database is loaded after authentication (without prompt for master password)

When you choose **Lock Files** and afterwards **close password files** (ie. not the big unlock button), you still see the biometric prompt? This would be a bug.

See this example: The first time I choose 'Close password files' and no biometric import pops up, but I get to the home screen. The second time i choose 'Unlock' which immediately reopens the files with biometric prompt.

![2020-09-22 22-56-46 2020-09-22 22_57_54](https://user-images.githubusercontent.com/313066/93936947-356b5580-fd27-11ea-9493-6b6b4ac5eb12.gif)


> When opening both files at the same time this is indeed possible, but it is not that convenient when all passwords of all databases are available. I would prefer to have them separate. Is this possible?

Isn't it more convenient to just change the filter from the app drawer, instead of closing and reopening anyway?

**Comment by Trekky12** [comment]
Yes, when I *Close Password Files" the biometric prompt is shown again.

I just discovered the filter and this could be a possibility. However I have databases with many groups and it is not that easy to find the second database in the list. Is it possibile to fold the subgroups of a database so it is easier to find the database? Additionally it would be great when the filter is saved for the next app start.

**Comment by hpoul** [comment]
> Yes, when I **Close Password Files** the biometric prompt is shown again.

this sounds like a bug. can you try one thing maybe: In the app overflow menu -> preferences ... in there is another app overflow menu 😅️ with *Clear Biometric Storage* - can you try that one? this one completely deletes anything stored by AuthPass in the keychain/keystore. And then try the whole Open -> Lock -> Close again if it works this time.

Also, which platform, device, OS version are you using?

> 
> I just discovered the filter and this could be a possibility. However I have databases with many groups and it is not that easy to find the second database in the list. Is it possibile to fold the subgroups of a database so it is easier to find the database? Additionally it would be great when the filter is saved for the next app start.

You mean the root tree nodes should be collapsed, not expanded? Right now that's not really easy todo .. I'll have to think about that.. sounds like a power user feature to me :-) Before doing tree collapse/expand .. i could imagine either:

1. Something explicitly for switching between open files in the 'filter' menu
2. or a "Save filter as ..." which is then visible in the 'filter' menu

I think 2 would be pretty powerful and easy to implement.. any thoughts? this would also lead to your second idea of persisting the active filter..

**Comment by Trekky12** [comment]
Unfortunately clearing the biometric storage does not change this behavior. I have a Huawei P10 with Android 9.

I think both of your ideas are a great way to solve this usecase. I would personally prefer the first method where you can easily switch between databases but when the last saved filter is always applied on opening this would be a very powerful feature 👍
